### DISCUSSION TITLE
Finding attractors

### DISCUSSION THREAD

**Discussion by yzongy** [hklarner/PyBoolNet#94] [Q&A]
A previous discussion solved the model checking problem, which was raised by `pyboolnet.attractors.compute_attractors(primes, "synchronous")` :
[https://github.com/hklarner/pyboolnet/issues/93]

Though this error was solved by manually calling the model_checking function, how do we get the attractors under synchronous update with a given Boolean network(example: [BottomStrain_GoldStandardGRNbased.txt](https://github.com/hklarner/pyboolnet/files/8292018/BottomStrain_GoldStandardGRNbased.txt)) ?

**Comment by hklarner** [comment]
- so what's the output of `compute_attractors(primes, "synchronous")`?
- by the way you can add a file name to save the result (useful if computation time is long)

**Reply by yzongy** [reply]
It came back with a model_checking exception:

> INFO attractors.completeness(..)
>     INFO created /var/folders/8t/qrykd8w57pggdyvkp5c5nl_c0000gn/T/pyboolnet_ti93zc77
>     INFO no
>     INFO working on minimal trapspace 1/32: 10011111110100101101010011011111111101111111110111111111111111111001000111001011111111110111110001111111100011011111111111100000111111001111111111111110111111110001111111111111001110011111100111111111111111111011111011111101101011111111111101101111101111111011111111111100011111111111111011101010111111111011101110011111000110111000111111111111101111100111110111011111101111101110111101101001001011111110101111100011001111111010111101111100001110111010100111111111000010100010101111101111101010111111101110010111111111110111111110111011101010111100111101111111111111101111101101110011000110111011011101111011111110111010111111111111001110110111111111011110001111100111111110101111111101111110011011101100101111101001111101110001111110101111111110111111011110101111110101000110111001110001101101111010001011111111110110111011111111001101101110100110000111111111101101001111111111111110011011111110110111111101111101101001111100111011111010110111011111000010111110111111101111110010110011100011111100010111011111111111001011101001111110111010110111111111111101111111111111101100111101111111010011111111011110111111111111101111111111110111111011100111011010011111110111111111111111111110011111011111110010110101011111001110111111111111100111111111011011001111111110110011111111110111111110100110111101111110100011100100000111101111111110111110110111111011011101011011110110111011101101111111110011101111110101011100001111110010011111111111111111101011000101110110111111111011110111010111011111011111110011111111101111011
>     INFO attractors.univocality(..)
>     INFO yes
>     INFO attractors.faithfulness(..)
>     INFO yes
>     INFO attractors.find_attractor_state_by_randomwalk_and_ctl(..)
>     INFO find_attractor_state_by_randomwalk_and_ctl(..)
>     INFO len(primes)=1517, update=synchronous, length=15170, attempts=10
>     INFO trial 0
>     INFO start: 10011111110100101101010011011111111101111111110111111111111111111001000111001011111111110111110001111111100011011111111111100000111111001111111111111110111111110001111111111111001110011111100111111111111111111011111011111101101011111111111101101111101111111011111111111100011111111111111011101010111111111011101110011111000110111000111111111111101111100111110111011111101111101110111101101001001011111110101111100011001111111010111101111100001110111010100111111111000010100010101111101111101010111111101110010111111111110111111110111011101010111100111101111111111111101111101101110011000110111011011101111011111110111010111111111111001110110111111111011110001111100111111110101111111101111110011011101100101111101001111101110001111110101111111110111111011110101111110101000110111001110001101101111010001011111111110110111011111111001101101110100110000111111111101101001111111111111110011011111110110111111101111101101001111100111011111010110111011111000010111110111111101111110010110011100011111100010111011111111111001011101001111110111010110111111111111101111111111111101100111101111111010011111111011110111111111111101111111111110111111011100111011010011111110111111111111111111110011111011111110010110101011111001110111111111111100111111111011011001111111110110011111111110111111110100110111101111110100011100100000111101111111110111110110111111011011101011011110110111011101101111111110011101111110101011100001111110010011111111111111111101011000101110110111111111011110111010111011111011111110011111111101111011
>     INFO end: 10011111110100101101010011011111111101111111110111111111111111111001000111001011111111110111110001111111100011011111111111100000111111001111111111111110111111110001111111111111001110011111100111111111111111111011111011111101101011111111111101101111101111111011111111111100011111111111111011101010111111111011101110011111000110111000111111111111101111100111110111011111101111101110111101101001001011111110101111100011001111111010111101111100001110111010100111111111000010100010101111101111101010111111101110010111111111110111111110111011101010111100111101111111111111101111101101110011000110111011011101111011111110111010111111111111001110110111111111011110001111100111111110101111111101111110011011101100101111101001111101110001111110101111111110111111011110101111110101000110111001110001101101111010001011111111110110111011111111001101101110100110000111111111101101001111111111111110011011111110110111111101111101101001111100111011111010110111011111000010111110111111101111110010110011100011111100010111011111111111001011101001111110111010110111111111111101111111111111101100111101111111010011111111011110111111111111101111111111110111111011100111011010011111110111111111111111111110011111011111110010110101011111001110111111111111100111111111011011001111111110110011111111110111111110100110111101111110100011100100000111101111111110111110110111111011011101011011110110111011101101111111110011101111110101011100001111110010011111111111111111101011000101110110111111111011110111010111011111011111110011111111101111011
>     INFO created /var/folders/8t/qrykd8w57pggdyvkp5c5nl_c0000gn/T/pyboolnet_8xy2f1ir
>     ERROR nusmv did not respond with return code 0: cmd=/Users/yzongy/.pyenv/versions/3.9.7/lib/python3.9/site-packages/pyboolnet/binaries/NuSMV-a/NuSMVa_mac64 -dcx -dynamic -df -coi /var/folders/8t/qrykd8w57pggdyvkp5c5nl_c0000gn/T/pyboolnet_8xy2f1ir, return_code=-11 output=*** This is NuSMV-A (compiled on Wed Sep 14 14:04:57 2016)
>     *** NuSMV-A is an extension of NuSMV 2.6.0
>     *** For more information on NuSMV-A see [http://github.com/hklarner/NuSMV-A](https://github.com/hklarner/NuSMV-A)
>     *** Enabled addons are: compass
>     *** For more information on NuSMV see http://nusmv.fbk.eu/
>     *** or email to [nusmv-users@list.fbk.eu](mailto:nusmv-users@list.fbk.eu).
>     *** Please report bugs to <Please report bugs to [nusmv-users@fbk.eu](mailto:nusmv-users@fbk.eu)>
> 
>     *** Copyright (c) 2010-2014, Fondazione Bruno Kessler
> 
>     *** This version of NuSMV is linked to the CUDD library version 2.4.1
>     *** Copyright (c) 1995-2004, Regents of the University of Colorado
> 
>     *** This version of NuSMV is linked to the MiniSat SAT solver.
>     *** See http://minisat.se/MiniSat.html
>     *** Copyright (c) 2003-2006, Niklas Een, Niklas Sorensson
>     *** Copyright (c) 2007-2010, Niklas Sorensson
> 
>     , error=b''
>     Traceback (most recent call last):
>     File "/Users/yzongy/BooleanNetworks/findTopStrainAttractors.py", line 31, in
>     pyboolnet.attractors.compute_attractors(primes, "synchronous")
>     File "/Users/yzongy/.pyenv/versions/3.9.7/lib/python3.9/site-packages/pyboolnet/attractors.py", line 109, in compute_attractors
>     state = find_attractor_state_by_randomwalk_and_ctl(primes=primes, update=update, initial_state=mints)
>     File "/Users/yzongy/.pyenv/versions/3.9.7/lib/python3.9/site-packages/pyboolnet/attractors.py", line 242, in find_attractor_state_by_randomwalk_and_ctl
>     if model_checking(primes, update, init, spec):
>     File "/Users/yzongy/.pyenv/versions/3.9.7/lib/python3.9/site-packages/pyboolnet/model_checking.py", line 108, in model_checking
>     answer = model_checking_smv_file(
>     File "/Users/yzongy/.pyenv/versions/3.9.7/lib/python3.9/site-packages/pyboolnet/model_checking.py", line 170, in model_checking_smv_file
>     raise Exception
>     Exception

**Reply by hklarner** [reply]
I would try a different tool, for example this one:
- https://people.kth.se/~dubrova/bns.html

**Reply by hklarner** [reply]
you can use the function `primes2bns` to create a bns file.
